toolkit/bridge/primitives/src/lib.rs

```diff
@@ -0,0 +1,275 @@
+			Self::Inert => {},
+			Self::ActiveV1 { data } => {
+				inherent_data.put_data(INHERENT_IDENTIFIER, data)?;
+			},
+		}
+		Ok(())
+	}
+
+	async fn try_handle_error(
+		&self,
+		identifier: &InherentIdentifier,
+		mut error: &[u8],
+	) -> Option<Result<(), sp_inherents::Error>> {
+		if identifier == &INHERENT_IDENTIFIER {
+			let error = InherentError::decode(&mut error).ok()?;
+			Some(Err(sp_inherents::Error::Application(Box::from(error))))
+		} else {
+			None
+		}
+	}
+}
+
+sp_api::decl_runtime_apis! {
+	/// Runtime API used by [TokenBridgeInherentDataProvider]
+	#[api_version(1)]
+	pub trait TokenBridgeIDPRuntimeApi {
+		/// Returns the current version of the pallet, 1-based.
+		fn get_pallet_version() -> u32;
+		/// Returns the currenlty configured main chain scripts
+		fn get_main_chain_scripts() -> Option<MainChainScripts>;
+		/// Returns the currently configured transfer number limit
+		fn get_max_transfers_per_block() -> u32;
+		/// Returns last data checkpoint saved in the pallet
+		fn get_last_data_checkpoint() -> Option<BridgeDataCheckpoint>;
+	}
+}
+
+#[cfg(feature = "std")]
+impl<RecipientAddress: Encode + Send + Sync> TokenBridgeInherentDataProvider<RecipientAddress> {
+	/// Creates new [TokenBridgeInherentDataProvider]
+	///
+	/// This function is version-aware and will create [TokenBridgeInherentDataProvider] based on
+	/// the version reported by the pallet through [TokenBridgeIDPRuntimeApi].
+	pub async fn new<Block, T>(
+		client: &T,
+		parent_hash: Block::Hash,
+		current_mc_hash: McBlockHash,
+		data_source: &dyn TokenBridgeDataSource<RecipientAddress>,
+	) -> Result<Self, InherentDataCreationError>
+	where
+		Block: BlockT,
+		T: ProvideRuntimeApi<Block> + Send + Sync,
+		T::Api: TokenBridgeIDPRuntimeApi<Block>,
+	{
+		let api = client.runtime_api();
+
+		if !api.has_api::<dyn TokenBridgeIDPRuntimeApi<Block>>(parent_hash)? {
+			log::info!(
+				"💤 Skipping token bridge transfer observation. Pallet not detected in the runtime."
+			);
+			return Ok(Self::Inert);
+		}
+
+		match api.get_pallet_version(parent_hash)? {
+			1 => Self::new_v1(client, parent_hash, current_mc_hash, data_source).await,
+			unsupported_version => {
+				Err(InherentDataCreationError::UnsupportedPalletVersion(unsupported_version, 1))
+			},
+		}
+	}
+
+	/// Creates new [TokenBridgeInherentDataProvider::ActiveV1]
+	pub async fn new_v1<Block, T>(
+		client: &T,
+		parent_hash: Block::Hash,
+		current_mc_hash: McBlockHash,
+		data_source: &dyn TokenBridgeDataSource<RecipientAddress>,
+	) -> Result<Self, InherentDataCreationError>
+	where
+		Block: BlockT,
+		T: ProvideRuntimeApi<Block> + Send + Sync,
+		T::Api: TokenBridgeIDPRuntimeApi<Block>,
+	{
+		let api = client.runtime_api();
+		let max_transfers = api.get_max_transfers_per_block(parent_hash)?;
+		let last_checkpoint = api.get_last_data_checkpoint(parent_hash)?;
+
+		let (transfers, new_checkpoint) =
+			(data_source.get_transfers(last_checkpoint, max_transfers, current_mc_hash).await)
```

Comment:
Are parentheses around this line required? I expect some lint in post-merge action to fail because of them.